Introduce generic Callbacks to support running callbacks on executors

ramitkataria opened the pull request:
~~Introduce a generic Callback model that can be used for Deadline callbacks for now and hopefully the on_*_callbacks for Dags and Tasks in the future.~~

Introduce a generic Callback model to run callbacks on executor workers. To be used for Deadline callbacks and the on_*_callbacks for Dags and Tasks that run on the Dag Processor. Implementation of some of the functions will be added in another PR so that this PR can be focused on the DB model and the Workload definition.


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ramitkataria committed:
Introduce generic Callbacks

Introduce a generic Callback model that can be used for Deadline
callbacks for now and hopefully the on_*_callbacks for Dags and Tasks
in the future.

ramitkataria commented on airflow-core/src/airflow/executors/workloads.py:
```diff
@@ -84,6 +83,14 @@ def key(self) -> TaskInstanceKey:
         )
 
 
+class Callback(BaseModel):
```
Similar to TaskInstance defined above this in this file

ramitkataria commented on airflow-core/src/airflow/executors/workloads.py:
```diff
@@ -132,6 +139,51 @@ def make(
         )
 
 
+class ExecuteCallback(BaseWorkload):
```
Similar to ExecuteTask defined above. There's a lot of duplicate code here that we can refactor but for now, I'm just looking to align on the definition

ramitkataria commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,147 @@
+    EXECUTOR = "executor"
+
+
+class Callback(Base):
```
This is the focus of this PR and what I'm looking to align on for all Callbacks

ashb replied:
There is a lot of overlap (at least conceptually) between this and the existing db_callback_request model -- I think we should consider merging them if we can

ramitkataria replied:
I agree - I was thinking of adding a subclass for Dag Processor callbacks that includes the method `get_callback_request(self) -> CallbackRequest`?

ramitkataria replied:
They're merged now!

ramitkataria reviewed (commented):
I've highlighted the areas of this PR I want to focus on and get everyone working on AIP-92 to agree on. Everything else is just refactoring to make existing functionality work. I'll work on unit tests once we agree on how these should work

ramitkataria committed:
Add DB migration

ashb commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,147 @@
+        return TriggererCallback(callback_def)
+    if isinstance(callback_def, SyncCallback):
+        return ExecutorCallback(callback_def)
+    raise RuntimeError(f"Cannot handle Callback of type {type(callback_def)}")
```
cc @kaxil as we are trying to remove such imports in core as part of #51553

kaxil replied:
oh yes please, let's avoid any new coupling i.e. not need to import sdk in Core. See if you can serialize and deserialize it instead

ramitkataria replied:
I see - wouldn't that just mean creating a copy of the TaskSDK definition to deserialize to? Which class do you think is the best example to follow for this kind of decoupling?

kaxil replied:
https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/models/db_callback_request.py#L33-L48 or the recent HITL PRs

ramitkataria replied:
I realized I was just checking the type name so I removed the imports without significant changes in implementation

ramitkataria committed:
Revert the last 2 commits

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table

ramitkataria committed:
Introduce generic Callbacks

Introduce a generic Callback model that can be used for Deadline
callbacks and the on_*_callbacks for Dags and Tasks that run on the
Dag Processor. Implementation of some of the functions will be added in
another PR so that this PR can be focused on the DB model.

ramitkataria commented on airflow-core/src/airflow/migrations/versions/0091_3_2_0_restructure_callback_table.py:
```diff
@@ -0,0 +1,100 @@
+    """
+    # Drop all rows. this will only affect users who have incomplete Dag Processor callbacks during migration.
+    op.execute("DELETE FROM callback_request")
+    # TODO: migrate existing rows for pending DagProcessor callbacks
```
I looked into this but generating UUIDs of the correct type in SQL seems a bit complex so I think it's best to add that in another PR

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table

ramitkataria committed:
Refactor workloads to reduce code duplication

ramitkataria committed:
Make fetch_method required in ExecutorCallback constructor

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table

ramitkataria committed:
Minor change of order in autogenerated edge3 api yaml because of refactoring Execute Task

ramitkataria commented on providers/edge3/src/airflow/providers/edge3/openapi/v2-edge-generated.yaml:
```diff
@@ -919,6 +917,8 @@ components:
           - type: string
           - type: 'null'
           title: Log Path
+        ti:
```
All the changes in this provider are because of a change in the order of the fields in ExecuteTask? The `ti` field was moved to the end because of a refactor to abstract common code to a super class

potiuk replied:
I see. I though tit is not relevant change. Thanks for the comment.

ramitkataria commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+        self.data = callback_def.serialize()
+
+
+class DagProcessorCallback(Callback):
```
This is moved here from DbCallbackRequest

ferruzzi commented on airflow-core/src/airflow/executors/workloads.py:
```diff
@@ -97,6 +107,13 @@ class ExecuteTask(BaseWorkload):
+    """Execute the given Task."""
+
+    ti: TaskInstance
+    """The TaskInstance to execute"""
```
Feel free to just ignore this nit-pick:  Maybe don't need this comment, or make it a # comment and add it to the line above?  Seems odd here.

ramitkataria replied:
Removed it - it got there because I copied it from the previous definition of ExecuteTask

ferruzzi commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+class CallbackFetchMethod(str, Enum):
+    """Methods used to fetch callback at runtime."""
+
+    # For when Dag Processor callbacks (on_success_callback/on_failure_callback) once they get moved to executors
```
Is "once they get moved" intentional or a typo?  I'm note sure I understand the comment the way it is written.

ramitkataria replied:
Fixed!

ferruzzi reviewed (approved):
Couple of nits around comments, but otherwise I think it looks good.

jason810496 commented on airflow-core/src/airflow/executors/workloads.py:
```diff
@@ -43,6 +43,10 @@ class BaseWorkload(BaseModel):
     """The identity token for this workload"""
 
+    @staticmethod
+    def generate_token(sub_id: str, generator: JWTGenerator | None = None):
```
```suggestion
    def generate_token(sub_id: str, generator: JWTGenerator | None = None) -> str:
```

jason810496 commented on airflow-core/src/airflow/executors/workloads.py:
```diff
@@ -119,14 +135,45 @@ def make(
+        generator: JWTGenerator | None = None,
+        bundle_info: BundleInfo | None = None,
+    ) -> ExecuteCallback:
+        from pathlib import Path
```
```suggestion

```

It seems we have already imported at top level. Same as line 126

jason810496 commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+# under the License.
+from __future__ import annotations
+
+import logging
```
Perhaps we should use structlog instead of logging for new logger.

jason810496 commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+
+    __mapper_args__ = {"polymorphic_identity": CallbackType.EXECUTOR}
+
+    def __init__(self, callback_def, fetch_method: CallbackFetchMethod, **kwargs):
```
Do we need to align the constructor with `TriggererCallback` ?

ramitkataria replied:
This is intentional because triggerer can only use `CallbackFetchMethod.IMPORT_PATH` since it doesn't have access to the Dag bag

jason810496 commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+        self.state = CallbackState.QUEUED
+
+    @staticmethod
+    def create_from_sdk_def(callback_def) -> Callback:
```
```suggestion
    def create_from_sdk_def(callback_def: AsyncCallback | SyncCallback) -> Callback:
```

ramitkataria replied:
I had them earlier but removed after discussion here: https://github.com/apache/airflow/pull/54796#discussion_r2304116137

ramitkataria replied:
I made another attempt to improve the typing here without using any SDK imports. This time, I used Protocol classes. Let me know what you think!

jason810496 replied:
I see, thanks for point out previous discussion.
Nice, using Protocol is a good idea.

jason810496 commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+
+    __mapper_args__ = {"polymorphic_identity": CallbackType.TRIGGERER}
+
+    def __init__(self, callback_def, **kwargs):
```
Same here for the type annotation

jason810496 commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,177 @@
+        self.priority_weight = priority_weight
+        self.created_at = timezone.utcnow()
+
+    def queue(self):
+        self.state = CallbackState.QUEUED
```
Do we need to explicitly commit or flush the session to persist the new state immediately?

ramitkataria replied:
I was thinking that I would do that where I call this function. This is more for observability than functionality as of now (maybe it could change at some point in the future). But if you think we should, pass in the session and add/flush right away, I could make that change

jason810496 replied:
Both works with me, I was just curious about the functionality of this method.
Thanks for the clarification, if observability is the case, then the current method LGTM.

jason810496 reviewed (commented):
Nice! Thanks for the PR!
Feel free to add me to the reviewer of refactoring Callback for further PRs.

ramitkataria committed:
PR fixes

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table

ramitkataria committed:
Make fetch method non nullable

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table

potiuk commented on airflow-core/src/airflow/models/callback.py:
```diff
@@ -0,0 +1,207 @@
+        self.data = callback_def.serialize()
+
+    def queue(self):
+        # TODO: queue the trigger
```
Is it still TODO: ?

potiuk replied:
(IF so - I think we need to descrihbe what's missing and what are consequences if we do not complete it before merge)

ramitkataria replied:
Yes, I've intentionally left that blank because implementing it would require refactoring the deadlines-related changes in the triggerer and trigger model to be refactored to use this class instead. I'm planning to do that (and other necessary deadline related changes) in a followup PR so this PR doesn't get larger than it already is. This PR will have no changes in functionality because this class is not being used anywhere right now for the user so no consequences.

potiuk reviewed (approved):
This looks good as the "foundational" change for the generic callbacks. I like the idea of having explicit callback and callback types defined as "types" of workload - there are certain properties of callbacks that make it "different" than regular tasks that makes such separation very much needed.

I guess it could be good if final review from @ashb @kaxil and @amoghrajesh are done (also because of potential impact on task-sdk work).

ramitkataria committed:
Merge branch 'main' into ramitkataria/callback-table
